---
title: "Genome Analysis - Final Project"
author: "Logan Langholz"
date: "May 4, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```

## Overview

  The following was developed to explore genomic GC content as it relates to gene length.  Using a combination of pybedtools and R, sequences from the coding regions of genes across all chromosomes from the hg19 reference genome were assesed for %GC content.

## Hypothesis
  Because the translational stop codons (TAG, TAA, and TGA) feature less GC content, coding regions with a higher GC% will tend to be longer in length.
  
## Methods
####Python Script:    
  The front end portion of this task was accomplished using the pybedtools suite within a python script.  The output of the scritp was a .tsv file that contained chromosome, start interval, ending interval, gene name, gene length, and %GC content for each gene in hg19.

```{python, eval=FALSE}
#! usr/bin/env python

########################################################################
'''Creates table for all hg19 genes that contain:
1) Chromosome name 2) start 3) end 4)gene name 5) gene length 6) GC%'''
########################################################################

from __future__ import absolute_import, division, print_function
import pybedtools
import ipdb

#---------------------
# hg19 genes bed file:
test_file='/Users/loganlangholz/Documents/class_files/Graduate/Spring_2016/Genome_analysis/Projects/data-sets/bed/test.genes.hg19.bed.gz'
filename='/Users/loganlangholz/Documents/class_files/Graduate/Spring_2016/Genome_analysis/Projects/data-sets/bed/genes.hg19.bed.gz'

test_genes = pybedtools.BedTool(test_file)
genes = pybedtools.BedTool(filename)

#-----------------------------------------------------

# fxn to get GC content from chr and interval of record in bed file:
def GC_content(chrom, start, end):

    fasta_file = '/Volumes/My Book/Genome_Analysis-Final_Project/%s.fa' %chrom
    fasta = pybedtools.BedTool(fasta_file)

    bed = pybedtools.BedTool('%s\t%s\t%s' %(chrom, start, end), \
    from_string = True)

    nuc_content = bed.nucleotide_content(fasta)
    GC = float(nuc_content[0][4]) #GC% is index4 of the 1st and only feature

    return GC

#-------------------------------------------------------
#Create tsv file with Chrom, Gene Name, Gene Length:

fw = open('tbl_chr-genes.tsv', 'w')

for record in genes:

    chrom = str(record.chrom)

    length = record.length
    name = str(record.name)
    start = int(record.start)
    end = int(record.end)
    GC_percent = GC_content(chrom, record.start, record.end)

    fw.write('%s\t%d\t%d\t%s\t%d\t%f\n' %(chrom, start, end, name, length, GC_percent))

fw.close()

```

#### Import Data Into R-Studio
  The next step was to import the generated .tsv file into R.  Using dplyr and tidyr verbs, the data table was joined with a file containing chromosomes and their lengths.  Each gene's relative position along the chromosome was calculated by dividing the starting interval by the chromosome length.  This data was used to explore any possible trends regarding GC content and relative chromosomal position.


```{r load_raw_data, message = FALSE}  
library(readr)
library(tidyr)
library(dplyr)
library(ggplot2)

#Table generated by python sript:
col_names <- c('chrom', 'start', 'stop', 'gene', 'gn_length', 'GC_content')
genes_length_gc <- read_tsv('tbl_chr-genes.tsv', col_names = col_names)

#Chromosome Lengths:
lengths_cols <- c('chrom', 'chr_length')
chrm_lengths <- read_tsv(
  '/Users/loganlangholz/Documents/class_files/Graduate/Spring_2016/Genome_analysis/Projects/data-sets/genome/hg19.genome', col_names = lengths_cols)

#Join tbls:
#new_tbl <- left_join(genes_length_gc, genes_bed, by = c('chrom', 'gene'))
master_tbl <- left_join(genes_length_gc, chrm_lengths, by = 'chrom')
master_tbl <- master_tbl %>% mutate(rel_position = start/chr_length)

#Threshold value for a 'large coding sequence':
lrg_gene_cutoff <- 2*(10^5)
```

## Generating Plots
  All plots were generated using the following code.  Data was split up in groups of four chromosomes at a time, to allow for better viewing. For the plots of GC content as a function of gene length, the trend line was modeled as `r 'y = log(x)'`.  
  
  For plots showing GC content as a function of relative chromosomal position, data was again split into groups of four chromosomes at a time.  All points were plotted in black, with a blue line representing the middle 50% GC line.  Data points that had a gene length above a certain threshold, `r as.integer(lrg_gene_cutoff)`bp, were overlayed and their color mapped as a funciton of log(gene length) on a gradient color scale (longest genes in yellow, shortest genes in red; all genes > `r as.integer(lrg_gene_cutoff)`). 
  
```{r}

#Chrs 1-4:
chrs <- c('chr1', 'chr2', 'chr3', 'chr4')
tbl_1_4 <- filter(master_tbl, master_tbl$chrom %in% chrs)

length_plot_1_4 <- ggplot(tbl_1_4, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x') 

rel_pos_plot_1_4 <-ggplot(tbl_1_4, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_1_4, tbl_1_4$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
```


```{r, echo=FALSE}
#Chrs 5-8:
chrs <- c('chr5', 'chr6', 'chr7', 'chr8')
tbl_5_8 <- filter(master_tbl, master_tbl$chrom %in% chrs)
length_plot_5_8 <- ggplot(tbl_5_8, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
rel_pos_plot_5_8 <-ggplot(tbl_5_8, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_5_8, tbl_5_8$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')

#Chrs 9-12:
chrs <- c('chr9', 'chr10', 'chr11', 'chr12')
tbl_9_12 <- filter(master_tbl, master_tbl$chrom %in% chrs)
length_plot_9_12 <- ggplot(tbl_9_12, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
rel_pos_plot_9_12 <-ggplot(tbl_9_12, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_9_12, tbl_9_12$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')

#Chrs 13-16:
chrs <- c('chr13', 'chr14', 'chr15', 'chr16')
tbl_13_16 <- filter(master_tbl, master_tbl$chrom %in% chrs)
length_plot_13_16 <- ggplot(tbl_13_16, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
rel_pos_plot_13_16 <-ggplot(tbl_13_16, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_13_16, tbl_13_16$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')

#Chrs 17-20:
chrs <- c('chr17', 'chr18', 'chr19', 'chr20')
tbl_17_20 <- filter(master_tbl, master_tbl$chrom %in% chrs)
length_plot_17_20 <- ggplot(tbl_17_20, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
rel_pos_plot_17_20 <-ggplot(tbl_17_20, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_17_20, tbl_17_20$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')

#Chrs 21-Y:
chrs <- c('chr21', 'chr22', 'chrX', 'chrY')
tbl_21_Y <- filter(master_tbl, master_tbl$chrom %in% chrs)
length_plot_21_Y <- ggplot(tbl_21_Y, aes(gn_length, GC_content)) + 
  geom_point(size = 1) +
  geom_smooth(method = 'lm', formula = y ~ log(x)) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
rel_pos_plot_21_Y <-ggplot(tbl_21_Y, aes(rel_position, GC_content)) + 
  geom_point( size = 1, alpha = 1/5) +
  geom_point(data = filter(tbl_21_Y, tbl_21_Y$gn_length > lrg_gene_cutoff), aes(color = log(gn_length)), size = .75) +
  scale_color_gradientn( sprintf("log(gn_length > %dbp)", lrg_gene_cutoff), colors = grDevices::heat.colors(5) ) +
  geom_segment(aes(x=0, y=0.5, xend = 1, yend = 0.5), color = 'blue', size = 0.25) +
  facet_wrap(~chrom, ncol = 2, scales = 'free_x')
```
  
  

## Plots
####Correlation of %GC and Length:
  
```{r, echo = FALSE}
length_plot_1_4
length_plot_5_8
length_plot_9_12
length_plot_13_16
length_plot_17_20
length_plot_21_Y
```

####Relationship between %GC and relative chromosomal position:

```{r, echo = FALSE}  
rel_pos_plot_1_4
rel_pos_plot_5_8
rel_pos_plot_9_12
rel_pos_plot_13_16
rel_pos_plot_17_20
rel_pos_plot_21_Y  
```  

## Observations
  The trend of the data seems to be in opposition of the original hypothesis.  GC content appears to decrease as the coding sequence length increases.  This trend appears to asymptotically approach a value of roughly 40% GC in all chromosomes.  Looking at the plots of GC content and relative chromosomal position, an overwhelming majority of coding sequences above the established threshold fall below the 50% GC line, indicating some sort of advantageous connetion between longer coding sequences and lower GC content.  